### PR TITLE
[7184] api v1_0 setup

### PR DESCRIPTION
**Pull request by d-a-v-e**
### Context

Introduced a service to automate the generation of new versioned API files based on existing ones, ensuring consistent structure and inheritance.

### Changes proposed in this pull request

- Implemented `ApiVersionGenerator` service to create new versioned API files
- Ran the generator to create new versioned API 
- Added `SpecVersionGenerator` to copy and modify specs for specified new version
- Ran the generator to create new versioned Specs for the API
- duplicated the openAPI specs and done a find/replace for v0.1
- duplicated the reference docs and done a find/replace for v0.1 and removed notes around draft version of the api

### Guidance to review

- Review the `ApiVersionGenerator` and `SpecVersionGenerator `
- Verify the new versioned API  and spec files generated correctly match the expected structure
- Ensure all tests pass successfully (they do for me!)
- There is some new (proposed) guidance for cutting a new [API version in the docs here](https://github.com/DFE-Digital/register-trainee-teachers/blob/364aa17bbb598d314d733794610592a35ec26fea/docs/api.md), which is worth reviewing 

### Notes
There is also a new rake folder for API related tasks, I've put in calls to both the new generators in here

```ruby
# lib/tasks/api.rake

namespace :api do
  desc "Generate new versioned API files"
  task :generate_new_version, %i[old_version new_version] => :environment do |_t, args|
    old_version = args[:old_version] || "v0.1"
    new_version = args[:new_version] || "v1.0"

    ApiVersionGenerator.call(old_version:, new_version:)
  end

  desc "Generate new versioned spec files"
  task :generate_new_spec_version, %i[old_version new_version] => :environment do |_t, args|
    old_version = args[:old_version] || "v0.1"
    new_version = args[:new_version] || "v1.0"

    SpecVersionGenerator.call(old_version:, new_version:)
  end
end
```

### CODE REVIEW

**Comment by d-a-v-e** [comment]
[Set up core functionality for API v1.0](https://trello.com/c/Ux6fsXAw/7184-set-up-core-functionality-for-api-v10)

File: app/controllers/api_docs/base_controller.rb [hunk 1]
```diff
@@ -3,7 +3,7 @@
 module ApiDocs
   class BaseController < ::ApplicationController
     VERSIONS = %w[v0.1 v1.0].freeze
-    CURRENT_VERSION = "v0.1"
+    CURRENT_VERSION = "v1.0"
```
[the remark this answers is not in the document]
Comment (defong) [reply app/controllers/api_docs/base_controller.rb:6]: Yeah I guess so, maybe also move it to both
`VERSIONS` and `CURRENT_VERSION` to the settings to allow for toggling between `ENV`

File: spec/models/api/v1_0/degree_attributes_spec.rb [hunk 2]
```diff
@@ -0,0 +1,62 @@
+# frozen_string_literal: true
+
+require "rails_helper"
+
+RSpec.describe Api::V10::DegreeAttributes do
```
Comment (defong) [inline spec/models/api/v1_0/degree_attributes_spec.rb:5]: Not too sure with duplicating the tests seems too much, causing it to be fragile...

Maybe we should extract it so we have some reuse value, a shared example or something.

If we get to x version and nothing was changed then we have x copy of the same lines of tests.

**Review by defong — COMMENTED** [review]
Looking good, just need a think about those tests.

File: app/controllers/api_docs/base_controller.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (d-a-v-e) [reply app/controllers/api_docs/base_controller.rb]: version is now retrieved from the settings

**Review by tomtrentham — APPROVED** [review]
This looks great 🚀

**Pull request merged**